src/URIs.jl

```diff
@@ -453,31 +455,54 @@ normpath(url::URI) =
     URI(scheme=url.scheme, userinfo=url.userinfo, host=url.host, port=url.port,
         path=normpath(url.path), query=url.query, fragment=url.fragment)
 
+# normpath helper functions
+_tail(s, prefix) = last(s, length(s) - length(prefix))
+function _pop_segment(buf)
+    last_slash = findlast('/', buf)
+    (last_slash === nothing) ? "" : first(buf, last_slash-1)
```

Comment:
```suggestion
    (last_slash === nothing) ? "" : buf[1:prevind(buf, last_slash)]
```

This will fix incorrect behaviour when using non-ASCII characters, e.g.

```
@test _pop_segment("a/b") == "a"
@test _pop_segment("🌟/🌟") == "🌟"
```

We should also add a couple of tests that use non-ASCII characters.

Reply:
RFC 3986 says that such characters should have been percent-encoded during construction, so this edge case shouldn't occur. 

```js
// JS
> new URL("https://a/🌟/e").pathname
"/%F0%9F%8C%9F/e"
```

```julia
julia> URI("https://a/🌟/e").path
"/🌟/e"
```

Let's make a new issue and go from there?

Reply:
I've gone ahead and incorporated your suggestion since it seems to be the better way of doing this in any case, but you're right -- URIs, as specified by RFC 3986, are only supposed to include US-ASCII characters. If we wanted to support Unicode we'd have to implement [RFC 3987](https://datatracker.ietf.org/doc/html/rfc3987), which specifies Internationalized Resource Identifiers (IRIs).